src/ignorer.rs

```diff
@@ -0,0 +1,71 @@
+use {
+    anyhow::{Context, Result},
+    git_repository::{
+        self as git,
+        prelude::FindExt,
+        Repository,
+    },
+    std::path::{Path, PathBuf},
+};
+
+/// An object able to tell whether a file is excluded
+/// by gitignore rules
+pub struct Ignorer {
+    repo: Repository,
+}
+
+impl Ignorer {
+
+    /// root_path is assumed to exist and be a directory
+    pub(crate) fn new(root_path: &Path) -> Result<Self> {
+        let repo = git::discover(root_path)?;
+        Ok(Self { repo })
+    }
+
+    /// Tell whether the given path is excluded according to
+    /// either the global gitignore rules or the ones of the repository
+    pub fn excludes(&mut self, file_path: &Path) -> Result<bool> {
```

Comment:
I think as it stands, each path will trigger reading all gitignore files. They are indeed held in the `excludes()` data structure and ideally it is kept. I see how this isn't possible right now, and believe that the current reference is likely a premature optimization rather than a necessity. This will change for sure - and it's done in the latest main, which should greatly improve performance as the cache can actually reuse state if it's kept around.

Reply:
> the current reference is likely a premature optimization rather than a necessity

I feel you. I fell in the same trap in my first libs too, and it's a pain to fix.

TBH checking whether a file is excluded takes less than 1 ms right now and that's fine for bacon.

Reply:
I will think about this - the number one usage of lifetimes in structs is _platforms_, they add some data around an operation and perform it, keeping a reference to their originating `Repository`. As long as these are basically free, I think they can be created on demand with the `Repository` being cloned to where it is needed - that's the intent.

If they aren't free though, like the `Cache` here, I think it's good advice to rather clone the `Repository` into it to make it standalone, or do whatever else it takes. I think some useful rule emerges from this experience and I will put it into words in `DEVELOPMENT.md` to make it official.

Reply:
BTW @Byron I don't know if you may find this interesting, but I've also implemented a gitignoring stack (stacking the parsed gitignore files as I imagine you do): https://github.com/Canop/broot/blob/main/src/git/ignore.rs#L155

This was done for broot with very specific performance concerns (breathfirst tree diving).
The reasons I didn't take that for bacon were
1. I wanted to try gitoxide for other programs of mine
2. I didn't want to implement myself looking for the global gitignore rules (I use git2 in broot for that)

Reply:
Thanks for sharing! I love the perceived simplicity of the git-ignore implementation, it fits in 240 lines after all :)!

With the `Cache` type unchained from the lifetime in `main` you would now be able to reuse it for each lookup and that should yield much better performance.

From a correctness point of view, it's probably (hopefully)  a good idea to use `gitoxide` even if the performance is _just_ similar, as I tried my best to validate the implementation against git with many many test cases. Of course I hope you won't take my word for it and validate it yourself, `gitoxide` strives to yield the same results as `git`.

Thus, I hope you will end up using `gitoxide` in more of your projects and if there is anything preventing that, I'd love to know to get a chance to fix it.

Reply:
gitignoring is only part of using git, and I definitely don't want to expand into building a general git crate while there's already an ambitious project, so it's my clear intention to try and use gitoxide ;)